connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -307,15 +307,25 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
     }
 
     private void processTargetStateChanges(Set<String> connectorTargetStateChanges) {
-        if (!connectorTargetStateChanges.isEmpty()) {
-            for (String connector : connectorTargetStateChanges) {
-                if (worker.connectorNames().contains(connector)) {
-                    TargetState targetState = configState.targetState(connector);
-                    worker.setTargetState(connector, targetState);
-                    if (targetState == TargetState.STARTED)
-                        reconfigureConnectorTasksWithRetry(connector);
-                }
+        for (String connector : connectorTargetStateChanges) {
+            TargetState targetState = configState.targetState(connector);
+            if (targetState == null) {
+                // The target state was deleted. This should follow only after deletion of the
+                // connector's configuration itself, so raise an error if otherwise.
+                if (configState.connectors().contains(connector))
+                    throw new ConnectException("Unexpected removal of target state for connector " + connector);
```

Comment:
Yeah, you're right. I was thinking that because we always write the config tombstone first, it would be possible to see a target state with no connector config, but not the other way around. However, we currently don't write the target state on connector creation, so it's definitely possible to see this case. One option to fix this would be to write the initial target state of the connector prior to its configuration when creating the connector. Either that or we can just log a message here.